perf(semantic): reduce match arms in `check_binding_identifier`

overlookmotel opened the pull request:
Small optimization.

1. First arm match on `VariableDeclarator`, instead of `VariableDeclaration` - requires 1 less iteration to get to.
2. Remove match for `Program`. There's no point as iterator does that check on each turn of the loop anyway.

Boshen commented:
### Merge activity

* **Jul 16, 1:18 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(semantic): reduce match arms in `check_binding_identifier` (#12304)

Small optimization.

1. First arm match on `VariableDeclarator`, instead of `VariableDeclaration` - requires 1 less iteration to get to.
2. Remove match for `Program`. There's no point as iterator does that check on each turn of the loop anyway.
